### DISCUSSION TITLE
field 的 value 发生改变时应该进行 validate

### DISCUSSION THREAD

**Discussion by liuweiGL** [alibaba/formily#1449] [Q&A 2.x] [answered]
## 当前行为
现在的 validate 需要根据校验的类型来调用 field 对应的方法，比如 onInput，onBlur 等等。但是使用 field.setValue 跟 form.setValues 等方法时不会去触发一次 validate。

## 前提
A 字段设置了 required 校验

## 场景：onInput
1. 用户先填写 A 字段，然后清空，这时候出现必填校验。
2. 用户填写 B 字段，B 字段对应的组件中使用 form.setValues 去更新了 A 字段，这时候 A 有值了却没清空错误信息。
---
B 组件需要使用 form.query(xxxx).take().onInput(value) 才能触发校验

## 场景：onBlur

1. 用户点击一下 A 字段的输入框，并不填写，这时候出现必填校验
2. 用户填写 B 字段，B 字段对应的组件中使用 form.setValues 去更新了 A 字段，这时候 A 有值了却没清空错误信息。
---
B 组件需要使用 form.query(xxxx).take().onBlue(value) 才能触发校验

## 问题

我觉得校验类型对标的是用户行为，即用户进行了什么操作就自动触发对应的校验。而在代码层面是不区分的，不能让开发者手动维护校验时机啊，这也太费劲了。

## 期望

希望 field 的 value 发生改变时会自动触发一次 validate

**Comment by janryWang** [comment]
这种黑盒逻辑会破坏模型的纯粹性，考虑一种场景，异步默认值，如果传进去的initialValues里，有部分数据是null，那要不要触发校验，不过你的出发点没问题，需要再想想最佳方案

**Reply by janryWang** [reply]
还有一种，用户是通过form.setValuesIn('xxx')去修改某个数据，而不是通过query查到字段单独修改value那种，那是不是得全量搜索字段路径匹配的字段去触发校验

**Comment by janryWang** [comment, marked as answer]
我的个人看法，onInput触发校验是确定性最强的逻辑，相反，值变化触发校验，只是部分场景需要，很多场景，完全不需要，这种会给代码带来巨大的历史包袱

**Comment by liuweiGL** [comment]
> 值变化触发校验，只是部分场景需要，很多场景，完全不需要

如果表单里面存在联动，就普通的表单而言都需要这个特性吧，不然开发者手动维护每个字段的校验时机还是很繁琐的，写多了感觉回到了 jquery 风格。

**Comment by liuweiGL** [comment]
## 校验时机
我认为值变化校验最主要的目的是校验存在错误信息的字段，相等于复查一下。把有校验错误的字段再校验一次，而校验没有错误的字段是不需要复查的。

## 功能添加方式

是否能像 validateFirst 一样，添加一个 autoClearErrors 属性，当值变化时自动校验一次已发生错误的字段以便自动清空错误信息。

**Comment by janryWang** [comment]
最新代码已支持该功能，具体规则必须满足以下所有条件才会触发校验：
- 值发生变化，且不是在onInput的阶段
- 该字段modified为true
- 值存在

**Reply by feyy** [reply]
目标field都没有被Focus过，为啥onFocus和onBlur要触发呢。setValuesIn本身就是程序的强制设置逻辑，要判断正确性也可以程序先判断的吧。
我现在有种通过ids批量填列表数据场景的情况，通过textarea可以批量提交一批id，然后blur的时候通过接口快速填充随后的一个List数据；
同时list数据也是可以手动单项修改和添加的。
这种场景我在list内部单项添加操作的时候，把id同步到批量操作的textarea里也会触发onblur的批量操作，然而此时我只是想同步一下表单项数据

**Reply by feyy** [reply]
目前修改$field.selfModified = false可以阻止触发onBlur

**Reply by janryWang** [reply]
你有没有看懂？onFocus/onBlur是传给组件的，组件什么时候调用是组件的事情，跟内核有啥关系

**Reply by feyy** [reply]
目前情况是form.setValuesIn触发了对应field的onBlur类型的validator，我通过修改$field.selfModified = false阻止了validator的触发。
上面不是说的这个事儿么？

**Reply by janryWang** [reply]
上面说的跟你的问题一点关系没有，建议单独开个discussion，带上codesandbox最小用例说明清楚你的问题
